src/snowflake/snowpark/ml/transformer.py

```diff
@@ -0,0 +1,422 @@
+            original_columns.append(self._output_cols[0])
+            df = df.with_column(self._output_cols[0], onehot_array)
+            df = df.select([col(c) for c in original_columns])
+        else:
+            onehot_arrays = []
+            for fc, oc in zip(self._states_table_cols, self._output_cols):
+                if sparse:
+                    onehot_array = array_construct_compact(
+                        df[f"{fc}_count"], df[f"{fc}_encoder"], lit(1)
+                    )
+                else:
+                    onehot_array = iff(
+                        df[f"{fc}_encoder"].is_null(),
+                        df[f"{fc}_sparse"],
+                        df[f"{fc}_encoder"],
+                    )
+                onehot_arrays.append(onehot_array)
+                original_columns.append(oc)
+            df = df.with_columns(self._output_cols, onehot_arrays)
+            df = df.select([col(c) for c in original_columns])
+        return df
+
+
+class OrdinalEncoder(Transformer):
+    def __init__(
+        self,
+        input_cols: Optional[List[str]] = None,
+        output_cols: Optional[List[str]] = None,
+        category: Union[str, List[List[str]]] = "auto",
+    ):
+        super().__init__(input_cols, output_cols)
+        self._category = category
+        self._merge = False
+        self._states_table_cols = (
+            [f"states_{c}" for c in self._input_cols] if self._input_cols else None
+        )
+        self._encoder_table_cols = (
+            [f"states_{c}_encoder" for c in self._input_cols]
+            if self._input_cols
+            else None
+        )
+        self._category_table_name = {}
+
+    @Transformer.input_cols.setter
+    def input_cols(self, value: List[str]) -> None:
+        self._input_cols = value
+        self._states_table_cols = (
+            [f"states_{c}" for c in self._input_cols] if self._input_cols else None
+        )
+        self._encoder_table_cols = (
+            [f"states_{c}_encoder" for c in self._input_cols]
+            if self._input_cols
+            else None
+        )
+
+    def fit(self, df: DataFrame) -> "OrdinalEncoder":
+        if not self._input_cols:
+            raise ValueError("Input columns can not be None")
+
+        encoder_fit(self, df, "ordinal")
+        return self
+
+    def transform(
+        self,
+        df: DataFrame,
+    ) -> DataFrame:
+        """
+        The ordinal encoder will be "null" if there are new types in dataframe fed into transform compared to dataframe
+        fed into fit() function.
+        """
+        if self._output_cols is None:
+            raise ValueError("Output column can not be empty")
+        elif len(self._output_cols) == 1:
+            self._merge = True
+        elif len(self._input_cols) == len(self._output_cols):
+            self._merge = False
+        else:
+            raise ValueError(
+                "The number of output colum and input column does not match"
+            )
```

Comment:
I'm curious if it makes sense to extract this check into base class given it is called in different sub classes. Ditto for other `transform` checks.